apps/prairielearn/assets/scripts/instructorInstanceAdminBillingClient.ts

```diff
@@ -0,0 +1,30 @@
+import morphdom = require('morphdom');
+import { decodeData, onDocumentReady } from '@prairielearn/browser-utils';
+
+import { InstructorInstanceAdminBillingForm } from '../../src/ee/lib/billing/components/InstructorInstanceAdminBillingForm.html';
+import { type PlanName } from '../../src/ee/lib/billing/plans-types';
+
+onDocumentReady(() => {
+  const billingForm = document.querySelector<HTMLFormElement>('.js-billing-form');
+  const initialProps = decodeData('billing-form-data');
+
+  const studentBillingCheckbox = document.querySelector<HTMLInputElement>('#studentBillingEnabled');
+  const computeCheckbox = document.querySelector<HTMLInputElement>('#computeEnabled');
+
+  billingForm.addEventListener('change', () => {
+    const basicPlanEnabled = studentBillingCheckbox.checked;
+    const computePlanEnabled = computeCheckbox.checked;
+
+    const requiredPlans: PlanName[] = [];
+    if (basicPlanEnabled) requiredPlans.push('basic');
+    if (computePlanEnabled) requiredPlans.push('compute');
+
+    morphdom(
+      billingForm,
+      InstructorInstanceAdminBillingForm({
+        ...initialProps,
+        requiredPlans,
+      }).toString()
+    );
```

Comment:
An alternative to this: use HTMX to hit a server endpoint to compute the new markup?

Reply:
I think this is quite nice as is. I do worry a bit that we are essentially writing our own client-side framework, but maybe things like `morphdom` and `selector-observer` are actually the way to _avoid_ client-side frameworks? I don't know enough to have a good opinion about whether it would be cleaner to use React or Svelte or whatever. I _do_ like the fact that `morphdom` just does one simple thing and it's easy to understand with no magic.

Reply:
Yeah, I definitely would like to avoid our own bespoke client-side "framework" here. If we start seeing this pattern appear in more places, I'll take a second look at this.

The problem with React and friends is that they don't necessarily plug in nicely with server-rendered HTML that wasn't also generated with React/etc. The "right" way to use them is with a meta-framework that takes care of hydration on the client, shuttling data from the backend to the frontend, etc. We could always build our own little helpers to do this, including rendering on the backend and hydrating on the frontend?

Regardless, let's kick this can down the road until we actually have working student-pays 😂

Reply:
Totally agree on kicking this down the road!